clustered/common/src/main/java/org/ehcache/clustered/common/messages/ChainCodec.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.common.messages;
+
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+import org.ehcache.clustered.common.store.SequencedElement;
+import org.ehcache.clustered.common.store.Util;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ *
+ */
+public class ChainCodec {
+
+  private static final byte NON_SEQUNCED_CHAIN = 0;
+  private static final byte SEQUENCED_CHAIN = 1;
+  private static final byte SEQ_NUM_OFFSET = 8;
+  private static final byte ELEMENT_PAYLOAD_OFFSET = 4;
+
+  public static byte[] encode(Chain chain) {
+    ByteBuffer msg = null;
+    boolean firstIteration = true ;
+    for (Element element : chain) {
+      if (firstIteration) {
+        firstIteration = false;
+        ByteBuffer buffer = ByteBuffer.allocate(1);
+        if (element instanceof SequencedElement) {
+          buffer.put(SEQUENCED_CHAIN);
+        } else {
+          buffer.put(NON_SEQUNCED_CHAIN);
+        }
+        buffer.flip();
+        msg = combine(buffer, encodeElement(element));
```

Comment:
Encoding a `Chain` seems to require quite a few `ByteBuffer` allocations and copies which looks fairly expensive. I wonder if pre-iterating the `Chain` once to calculate the final `byte[]` size wouldn't be a lot cheaper despite the fact that the cain would need to be iterated twice.